wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/environment/server/CheEnvironmentEngine.java

```diff
@@ -0,0 +1,887 @@
+     *         whether machines from environment should be recovered or not
+     * @param messageConsumer
+     *         consumer of log messages from machines in the environment
+     * @return list of running machines of this environment
+     * @throws ServerException
+     *         if other error occurs
+     */
+    public List<Instance> start(String workspaceId,
+                                Environment env,
+                                boolean recover,
+                                MessageConsumer<MachineLogMessage> messageConsumer) throws ServerException,
+                                                                                           ConflictException {
+
+        // Create a new start queue with a dev machine in the queue head
+        List<MachineConfigImpl> startConfigs = env.getMachineConfigs()
+                                                  .stream()
+                                                  .map(MachineConfigImpl::new)
+                                                  .collect(Collectors.toList());
+        final MachineConfigImpl devCfg = removeFirstMatching(startConfigs, MachineConfig::isDev);
+        startConfigs.add(0, devCfg);
+
+        EnvironmentHolder environmentHolder = new EnvironmentHolder(new ArrayDeque<>(startConfigs),
+                                                                    new CopyOnWriteArrayList<>(),
+                                                                    messageConsumer,
+                                                                    EnvStatus.STARTING,
+                                                                    env.getName());
+
+        EnvironmentHolder returnedEnv;
+        try (StripedLocks.WriteLock lock = stripedLocks.acquireWriteLock(workspaceId)) {
+            returnedEnv = environments.putIfAbsent(workspaceId, environmentHolder);
+        }
+        if (returnedEnv != null) {
+            throw new ConflictException(format("Environment of workspace '%s' already exists", workspaceId));
+        }
+
+        startQueue(workspaceId, env.getName(), recover, messageConsumer);
+
+        List<Instance> machines;
+        try (StripedLocks.WriteLock lock = stripedLocks.acquireWriteLock(workspaceId)) {
+            environmentHolder = environments.get(workspaceId);
+            // possible only if environment was stopped during its start
+            if (environmentHolder == null) {
+                throw new ServerException("Environment start was interrupted by environment stopping");
+            }
+            machines = environmentHolder.machines;
+            environmentHolder.status = EnvStatus.RUNNING;
+        }
+        return new ArrayList<>(machines);
+    }
+
+    /**
+     * Stops running environment of specified workspace.
+     *
+     * @param workspaceId
+     *         ID of workspace that owns environment
+     * @throws EnvironmentNotRunningException
+     *         if environment is not found
```

Comment:
mb better to write "when environment is not running", because it confuses in situation when environment exists but its status e.g. 'Starting', wdyt?